Add AGENTS.md (#99)
Imageomics/catalog: AGENTS.md

# Agents Guide

This file provides guidance for AI coding agents (e.g., GitHub Copilot) working on this repository.

## Project Overview

This is a **template repository** for a web-based catalog of an organization's public code, datasets, models, and spaces. It fetches live data from the GitHub API and the Hugging Face API and renders a searchable, filterable catalog page as a static site via GitHub Pages. The default setup is for the Imageomics Organization's GitHub and Hugging Face products, and this Imageomics Catalog is live.

## Critical: Templated Design

> **Do not restructure the codebase.** This project is intentionally designed as a template where the only expected user customization is editing `public/config.yaml` at instantiation — and occasionally updating `ADDITIONAL_REPOS` or `ADDITIONAL_HF_REPOS` when non-org repos must be added. The rest of the code is intended to be stable and reusable across forks.

Reviewer: ```suggestion
**Do not restructure the codebase.** This project is intentionally designed as a template where the only expected user customization is editing `public/config.yaml` at instantiation — and occasionally updating `ADDITIONAL_REPOS` or `ADDITIONAL_HF_REPOS` when non-org repos must be added. The `public/config.yaml` file is intended to be stable and reusable across repositories utilizing this template; new functionality should not require changes to existing configuration settings.
```
@EmersonFras, what do you think of this change? Trying to avoid it's overreactive testing and make clear the template user needs.

Reviewer 2: I think this is a good addition.

Additionally you could split into 2 different points about what changes are not welcomed (i.e. total restructure of the project etc.) and what changes are welcomed by agents. There could be a scenario where an overly cautious agent might make no changes at all when one is needed as there is no positive guidance.

Small point too but I would note with this section too that we are not mentioning the file tag-groups.js as a possible file that would need customization.

> **Do not restructure the codebase.** This project is intentionally designed as a template where the only expected user customization is editing `public/config.yaml` at instantiation — and occasionally updating `ADDITIONAL_REPOS` or `ADDITIONAL_HF_REPOS` when non-org repos must be added. The `public/config.yaml` file is intended to be stable and reusable across repositories utilizing this template; new functionality should not require changes to existing configuration settings.

Reviewer: ```suggestion
**Do not restructure the codebase.** This project is intentionally designed as a template where the only expected user customization is editing `public/config.yaml` and `public/tag-groups.js` at instantiation — and occasionally updating `ADDITIONAL_REPOS` or `ADDITIONAL_HF_REPOS` when non-org repos must be added. 

The `public/config.yaml` file is intended to be stable and reusable across repositories utilizing this template; new functionality should not require changes to existing configuration settings. The `public/tag-groups.js` file is likely to be unique to each template user and should have tag groups added as needed (proper use is described in `docs/tag-grouping-process.md`); **do not delete tag groups**.

Source files (`src/`) and scripts (`scripts/`) can be updated as needed for new functionality as long as changes adhere to the templated design principles established here.
```

When resolving issues or implementing features:
- Make the **smallest change** that correctly fixes the problem.
- **Do not restructure, consolidate, or refactor** functions or files unless it is strictly necessary for a correct, efficient solution.
- A function doing multiple related things is intentional — do not split it up unless there is a clear, specific reason to do so.

> - When in doubt, ask for a plan before making changes.

Reviewer: ```suggestion
- When in doubt, ask for a plan before making changes.
- If asked to make a fix in a pull request review, provide the fix as one or more suggested changes instead of directly committing changes.
```

This matches the debugging philosophy documented in [`README.md`](README.md#note-on-debugging-with-gemini): *"without changing the structure of existing code more than necessary for efficient design."*

## Repository Layout

| Path | Purpose |
|---|---|
| `public/config.yaml` | **Primary customization point.** All org names, colors, API settings, and extra repos live here. |

> | `public/tag-groups.js` | Canonical tag aliases; maps raw tags to normalized display tags. |

Reviewer: ```suggestion
| `public/tag-groups.js` | **Secondary customization point.** Canonical tag aliases; maps raw tags to normalized display tags. |
```

- `tests/config.integration.test.js` — confirms `public/config.yaml` passes validation
- `tests/filterAndSort.test.js` — unit tests for `src/filterAndSort.js`
- `tests/filterNewAdditionalEntries.test.js` — unit tests for `src/filterNewAdditionalEntries.js`
- `tests/normalizeTag.test.js` — unit tests for `src/normalizeTag.js`

When adding a new utility to `src/`, add a corresponding test file. Do **not** add tests that make live network calls.

## Configuration

All runtime behavior is controlled by `public/config.yaml`. It is fetched at page load (not bundled), so changes take effect on the next page load without a rebuild. Key fields:

- `ORGANIZATION_NAME` — used in all API calls (must be lowercase)
- `API_BASE_URL` — Hugging Face API base (default: `https://huggingface.co/api/`)
- `PLATFORM` — code platform; only `github` is currently supported
- `ADDITIONAL_REPOS` — forked or external GitHub repos to include
- `ADDITIONAL_HF_REPOS` — Hugging Face repos from outside the org to include (each entry needs `repo` and `type`)

The validator in `src/validateConfig.js` is the authoritative list of required config fields and their shapes.

## Build & Dev

```console

> npm install       # install dependencies (Node 24 required)

Reviewer 2: Might want to point towards `package.json` or `.nvrmc` to get the node version required just so it doesn't become stale after time.

> npm install       # install dependencies (Node 24 required)

Reviewer: ```suggestion
npm install       # install dependencies (Node version in .nvmrc)
```

# Agents Guide

This file provides guidance for AI coding agents (e.g., GitHub Copilot) working on this repository.

## Project Overview

> This is a **template repository** for a web-based catalog of an organization's public code, datasets, models, and spaces. It fetches live data from the GitHub API and the Hugging Face API and renders a searchable, filterable catalog page as a static site via GitHub Pages. The default setup is for the Imageomics Organization's GitHub and Hugging Face products, and this Imageomics Catalog is live.

Reviewer 3: ```suggestion
This is a **template repository** for a web-based catalog of an organization's public code, datasets, models, and spaces. It fetches live data from the code repository platform (such as GitHub) API and the platform API for hosting datasets and models (Hugging Face) and renders a searchable, filterable catalog page as a static site via GitHub Pages. The default setup is for the Imageomics Organization's GitHub and Hugging Face products.
```

Reviewer 3: Was the _live_ or _this Imageomics Catalog_ meant to be linked to the live site? Also, note that in a fork this wording won't work well. Maybe more like this: _The default setup is for the Imageomics Organization's GitHub and Hugging Face products._ (I.e., it's not clear what the reference to the live site would add to the coding agent's ability to work better.)

Reviewer: I was thinking it should know that it is in fact run off this repository, but perhaps leaving the default setup information is sufficient since the deploy workflow is described later.

When resolving issues or implementing features:
- Make the **smallest change** that correctly fixes the problem.
- **Do not restructure, consolidate, or refactor** functions or files unless it is strictly necessary for a correct, efficient solution.
- A function doing multiple related things is intentional — do not split it up unless there is a clear, specific reason to do so.
- When in doubt, ask for a plan before making changes.
- If asked to make a fix in a pull request review, provide the fix as one or more suggested changes instead of directly committing changes.

This matches the debugging philosophy documented in [`README.md`](README.md#note-on-debugging-with-gemini): *"without changing the structure of existing code more than necessary for efficient design."*

## Repository Layout

| Path | Purpose |
|---|---|
| `public/config.yaml` | **Primary customization point.** All org names, colors, API settings, and extra repos live here. |
| `public/tag-groups.js` | **Secondary customization point.** Canonical tag aliases; maps raw tags to normalized display tags. |
| `index.html` | Static HTML shell; config values are applied dynamically. |

> | `style.css` | Custom styles; colors are set via CSS custom properties populated from `config.yaml`. |

Reviewer: Resolved for all tables in [320a947](https://github.com/Imageomics/catalog/pull/99/commits/320a947dc81ced20d9df4ba22b191b465afb2845).

Current test files:
- `tests/validateConfig.test.js` — unit tests for `src/validateConfig.js`
- `tests/config.integration.test.js` — confirms `public/config.yaml` passes validation
- `tests/filterAndSort.test.js` — unit tests for `src/filterAndSort.js`
- `tests/filterNewAdditionalEntries.test.js` — unit tests for `src/filterNewAdditionalEntries.js`
- `tests/normalizeTag.test.js` — unit tests for `src/normalizeTag.js`

When adding a new utility to `src/`, add a corresponding test file. Do **not** add tests that make live network calls.

## Configuration

All runtime behavior is controlled by `public/config.yaml`. It is fetched at page load (not bundled), so changes take effect on the next page load without a rebuild. Key fields:

- `ORGANIZATION_NAME` — used in all API calls (must be lowercase)
- `API_BASE_URL` — Hugging Face API base (default: `https://huggingface.co/api/`)

> - `PLATFORM` — code platform; only `github` is currently supported

Reviewer 3: ```suggestion
- `PLATFORM` — code repository platform; only `github` is currently supported
```
